File: selenium_testng_parallel_with_wdm/src/test/java/org/o2f/pages/GoogleSearchPage.java
package org.o2f.pages;

import org.openqa.selenium.Keys;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;

import java.util.List;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.containsString;

public class GoogleSearchPage extends BasePage{

    @FindBy(name = "q")
    private WebElement searchBox;

    @FindBy(name = "btnK")
    private List<WebElement> searchBtn;

    @FindBy(id = "L2AGLb")
    private WebElement agreeBtn;

    @FindBy(xpath = "//h3")
    private List<WebElement> resultsLinks;

    public GoogleSearchPage(WebDriver driver) {
        super(driver);
    }

    public void goToGoogle(){
        driver.get("https://www.google.com");
    }

    public void acceptConditions(){
        clickOnBtn(agreeBtn);
    }

    public void doSearch(String text){
        inputText(searchBox,text);
        clickOnKey(searchBox,Keys.ENTER);
    }

    public void validateFirstResultContainsSentence(String sentence){
        if (resultsLinks.size() > 0){
            String text = resultsLinks.get(0).getText();
            assertThat(text,containsString(sentence));
        }
    }
}

File: python-selenium-pure/pages/google_search_page.py
from selenium.webdriver import Keys
from selenium.webdriver.common.by import By

from config import settings
from pages.base_page import BasePage


class GoogleSearchPage(BasePage):

    search_box: By = (By.NAME, "q")
    search_btn: By = (By.NAME, "btnK") # this is a list of elements
    agree_btn: By = (By.ID, "L2AGLb")
    results_links: By = (By.XPATH, "//h3") # this is a list of elements

    def __init__(self, driver):
        super().__init__(driver)

    def open_google_page(self):
        self.goto_page(settings.google_base_url)

    def accept_conditions(self):
        self.click_on_btn(self.agree_btn)

    def perform_search(self, text: str):
        self.input_text(self.search_box,text)
        self.use_keyboard_key(self.search_box, Keys.ENTER)

    def validate_first_result_contains_sentence(self, sentence: str):
        elements = self.wait_until_elements_displayed(self.results_links)
        if len(elements) > 0:
            text = elements[0].text
            assert sentence in text
